### DISCUSSION TITLE
Intercept and modify WebSocket requests to inject bearer authentication token in header

### DISCUSSION THREAD

**Discussion by sanjaykulkarni04** [cefsharp/CefSharp#4011] [Q&A]
Hi,

Is there any way to intercept and modify WebSocket requests to inject bearer authentication tokens in the header?

Thanks,
Sanjay Kulkarni

**Comment by amaitland** [comment]
CEF does not currently support modification of websocket requests. Upstream support would need to be added.

**Reply by amaitland** [reply]
You can try a generic JavaScript based solution like https://github.com/skepticfx/wshook#easily-intercept-and-modify-websocket-requests-and-message-events

**Reply by sanjaykulkarni04** [reply]
> CEF does not currently support modification of websocket requests. Upstream support would need to be added.

Thank you @amaitland for responding. Can you please elaborate more on **_"Upstream support would need to be added"_** ?

**Reply by sanjaykulkarni04** [reply]
> You can try a generic JavaScript based solution like https://github.com/skepticfx/wshook#easily-intercept-and-modify-websocket-requests-and-message-events

In our system, Javascript doesn't have knowledge about authorization servers (IdentityServer 4) whereas the backend (C# component) is responsible for passing credentials to the identity server and getting the bearer token in response. After getting the JWT token we set it in HTTP Authorization header via HTTP request interceptor (CEF ResourceRequestHandler/IRequest). This works perfectly for HTTP requests.

Now in order to authenticate web socket requests how should we use this JavaScript-based solution? Can you please help me?

**Reply by amaitland** [reply]
> > CEF does not currently support modification of websocket requests. Upstream support would need to be added.
> 
> Thank you @amaitland for responding. Can you please elaborate more on **_"Upstream support would need to be added"_** ?

Please read first few paragraphs of  https://github.com/cefsharp/CefSharp/wiki/General-Usage for a better understanding of the project.

**Reply by sanjaykulkarni04** [reply]
> > You can try a generic JavaScript based solution like https://github.com/skepticfx/wshook#easily-intercept-and-modify-websocket-requests-and-message-events
> 
> In our system, Javascript doesn't have knowledge about authorization servers (IdentityServer 4) whereas the backend (C# component) is responsible for passing credentials to the identity server and getting the bearer token in response. After getting the JWT token we set it in HTTP Authorization header via HTTP request interceptor (CEF ResourceRequestHandler/IRequest). This works perfectly for HTTP requests.
> 
> Now in order to authenticate web socket requests how should we use this JavaScript-based solution? Can you please help me?

@amaitland Can you please suggest?

**Reply by sanjaykulkarni04** [reply]
> > > CEF does not currently support modification of websocket requests. Upstream support would need to be added.
> > 
> > 
> > Thank you @amaitland for responding. Can you please elaborate more on **_"Upstream support would need to be added"_** ?
> 
> Please read first few paragraphs of https://github.com/cefsharp/CefSharp/wiki/General-Usage for a better understanding of the project.

@amaitland Sure will go through it

**Reply by amaitland** [reply]
> Can you please suggest?

Not really. I've never had need to add a web socket headers.

See also https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=16376#p40095
Using a proxy server might be another option.

**Reply by sanjaykulkarni04** [reply]
Thank you @amaitland sorry for the delay in responding. Actually, CefServer is C++, not C#. Do you have any example/samplecode for C#?

**Reply by amaitland** [reply]
.net already has a websockets capable web server. Use that instead.

**Reply by sanjaykulkarni04** [reply]
ok. so it should be running in CEF or in a different process?
